FWCore/ServiceRegistry/interface/ActivityRegistry.h

```diff
@@ -28,11 +28,16 @@ to this file that go beyond the obvious cut and paste type of edits.
   1. The number at the end of the AR_WATCH_USING_METHOD_X macro definition
   is the number of function arguments. It will not compile if you use the
   wrong number there.
-  2. Use connect or connect_front depending on whether the callback function
+  2. Inside the watch function definition, choose either connect or
+  connect_front depending on whether the callback function
   should be called for different services in the order the Services were
-  constructed or in reverse order. Begin signals are usually forward and
-  End signals in reverse, but if the service does not depend on other services
-  and vice versa this does not matter.
+  constructed or in reverse order. If services don't depend on one another,
+  this choice does not matter (most services fall in this category, maybe
+  all). There is some inconsistency in how this has been implemented. Originally,
+  we intended usually with begin signals the services are called in forward
+  order and with end signals in reverse order. But a lot of these are implemented
+  with pre signals forward and post signals in reverse (not sure why or if
+  this was done intentionally or if it actually matters).
```

Comment:
I'd quickly think the pre-signals would be forward, and post-signals in reverse, but probably we should think more of what would make sense.

Reply:
After looking at them, I can see you are correct, most of them are implemented with Pre signals in forward order and Post signals in reverse order. The BeginJob and EndJob signals (transition level not module level) have begin signals forward and end signals in reverse. This is an implementation detail where we could spend a lot of time studying this and debating, where for most new transitions it probably does not matter one way or the other. For it to matter, services not only have to depend on each but that dependence also has to include a dependence on the ordering for a specific transition.